### ISSUE TITLE
Traces are enabled in the development environment polluting log files during flush

### ISSUE THREAD

**Issue by jarshwah** [DataDog/dd-trace-py#3288]
### Which version of dd-trace-py are you using?

0.51.0

### Which version of pip are you using?

21.1.1

### How can we reproduce your problem?

Decorate some functions with `ddtrace.tracer.wrap` and use `span.set_tags` within decorated functions. Ensure those functions are called in the development environment.

Because `ddtrace.tracer.enabled` defaults to True, whenever the agent attempts to flush to the stats collector, it emits errors into the log output, which can be very confusing for developers.

### What is the result that you get?

```
ERROR:ddtrace.internal.writer:failed to send traces to Datadog Agent at http://localhost:8126
Traceback (most recent call last):
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/tenacity/__init__.py", line 426, in __call__
    result = fn(*args, **kwargs)
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/ddtrace/internal/writer.py", line 356, in _send_payload
    response = self._put(payload, headers)
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/ddtrace/internal/writer.py", line 332, in _put
    conn.request("PUT", self._endpoint, data, headers)
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 950, in send
    self.connect()
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/http/client.py", line 921, in connect
    self.sock = self._create_connection(
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/socket.py", line 808, in create_connection
    raise err
  File "/Users/josh/.pyenv/versions/3.8.8/lib/python3.8/socket.py", line 796, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/ddtrace/internal/writer.py", line 458, in flush_queue
    self._retry_upload(self._send_payload, encoded, n_traces)
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/tenacity/__init__.py", line 423, in __call__
    do = self.iter(retry_state=retry_state)
  File "/Users/josh/.pyenv/versions/myproject/lib/python3.8/site-packages/tenacity/__init__.py", line 373, in iter
    six.raise_from(retry_exc, fut.exception())
  File "<string>", line 3, in raise_from
tenacity.RetryError: RetryError[<Future at 0x132521ac0 state=finished raised ConnectionRefusedError>]
```

### What is the result that you expected?

The tracing library does not emit events when the stats collector is not running, and therefore never emits errors.

### What's going on?

https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/tracer.py#L129 the ddtrace tracer is enabled by default. To switch it off an environment variable needs to be set (`DD_TRACE_ENABLED=0`) OR the default global tracer needs to be disabled `ddtrace.tracer.enabled = False`. Neither of these options are particularly nice though, because it's extremely difficult to know in which contexts the stats collector is actually running!

It might be easy enough to ask each of your developers to set an environment variable, or a localdev startup script to modify a global. But this doesn't translate to other environments (like test) where the agent may not be running.

Since the tracer should only be enabled when the collector is running (the agent has started the process) could a specific environment variable be set to communicate what the default should be?

For example, when the agent starts up, it injects a variable into the environment:

`DD_COLLECTOR_RUNNING=1`

Then when the tracer library is configuring itself, it could do the following:

```
COLLECTOR_RUNNING = asbool(get_env("collector", "running", default=False))
self.enabled = asbool(get_env("trace", "enabled", default=COLLECTOR_RUNNING))
```

So the default is still to be enabled when the collector is running, but false otherwise? Any explicit configuration of the enabled property is still respected.

**Comment by brettlangdon** [comment]
Hey @jarshwah thanks for taking the time to open this issue! This is definitely something we have seen before.

We opt for the current behavior, enabled by default, and always trying to connect to the agent because we want to avoid a scenario where you start your application in production expecting data to be collected and then it silently fails or disables itself because there is not an agent running (or network connection to another container is not setup properly, etc).

The best way for us to know if a trace agent is available is to try making an HTTP call to the agent.

We've discussed in the past that we could always disable after the first failure (oh no, no agent? ok we'll stop trying to send). However, the place where that becomes a problem is intermittent networking issues which can cause individual HTTP calls to fail or if the agent is ever restarted/updated and therefore a connection fails. If we did disable under this conditions then production applications could intermittently just stop sending data without a good way to providing that feedback to users.

We also enabled by default for ease of deployment/onboarding.

These are some of the factors which have led us to have the tracer enabled by default, and where a manual step is needed to tell the trace agent to not try and send data to the trace agent.


> For example, when the agent starts up, it injects a variable into the environment:

As far as the agent injecting an environment variable. This is a cool idea, but I am not sure exactly how/under what conditions it would work. Since the trace agent runs as either a separate process on the system or (generally these days) as a side car container. It likely won't be able to add the environment variable for all Python processes started. If it is running on the same host it could likely inject a global/default environment variable into `/etc/profile.d/` but I am unsure when that is/isn't used for every process/user.

If we relied on `DD_COLLECTOR_RUNNING` for the default value, this would mean changing the default for most users to off by default unless they are running under a scenario where the env variable could be injected.


One potential solution (if you are using `pytest`), instead of relying on setting env variables, is to use a `conftest.py` to set `ddtrace.tracer.enabled = False` that way the tracer is always disabled for your test suite.

Apologies, this ended up being a little more of a wall of text than I intended, but hopefully these thoughts/additional context help?

**Comment by jordan-carson** [comment]
First off - I am running an app using FlaskAPI, which is served via `uvicorn --host=0.0.0.0 --port=8050`. My team however has configured docker daemonsets within our EKS cluster. I am using `ddtrace-run python api.py etc` to run and trace the application. This results with the same error.

With that being said, I am running my work locally now with no environment really configured and I get the same error above. What do we need to configure to get around this, what is not yet configured our host and port for datadog?

**Comment by jarshwah** [comment]
Thanks for getting back to me @brettlangdon 

Your points regarding sidecar containers are good ones. I was only really considering the flow where the agent is responsible for starting the program, and therefore "injecting" environment variables really just means setting the environment before starting the child program. But if there are other common methods for running the agent/collector that doesn't involve the agent starting the program directly then my idea doesn't account for that at all.

> we want to avoid a scenario where you start your application in production expecting data to be collected and then it silently fails or disables itself because there is not an agent running

Understandable. I was hoping there might be a better way to know that the agent is definitely not running, but as discussed I see that's not really possible.

`conftest.py` will work for CI no problem. We also have developer setting files we can use to globally disable tracing. I was hoping to avoid the need there, or perhaps for a better way to silence/obscure the logging errors for environments that don't care.

**Comment by brettlangdon** [comment]
> I am running an app using FlaskAPI, which is served via `ddtrace-run uvicorn --host=0.0.0.0 --port=8050`. Served within an EKS cluster with datadog daemonsets running. We are serving the app but receive the same error as above.
> 
> We have done this within a separate docker container and receive the same error. What do we need to configure to get around this, is there an issue with the latest version? If running locally, what env vars are required to receive traces on datadog hq?

@jordan-carson would you be able to open a separate issue for this as it is unrelated to the original issue.

**Comment by brettlangdon** [comment]
@jarshwah 🤔 

The other approach is to disable logging for our library entirely.


For example:
```python
import logging

ddtrace_logger = logging.getLogger("ddtrace")
ddtrace_logger.propagate = False
ddtrace_logger.setLevel(logging.CRITICAL)  # or just make sure there isn't a handler configure for this logger
```

Untested, but something like that should work. I'd of course be worried that you might miss out on other important log messages so you could always do that configuration for `logging.getLogger("ddtrace.internal.writer")` to only ignore logs coming from that module.

**Comment by brettlangdon** [comment]
@jarshwah you should follow this thread as well: https://github.com/DataDog/dd-trace-py/issues/3306

**Comment by alysivji** [comment]
> The best way for us to know if a trace agent is available is to try making an HTTP call to the agent.

@brettlangdon Is there a health check endpoint we can hit? I tried a few different searches, but can't seem to find any documentation about an agent health check endpoint

For now I'm hitting the base agent URL (in my GKE: `http://datadog.default.svc.cluster.local:8126`) and assuming a 404 means the agent is running.

Thanks!

**Comment by brettlangdon** [comment]
> 

@alysivji there is a `/info` endpoint that has been around for "awhile" that might contain slightly more interesting information, but hitting `/` and expecting a `404` is probably ok as well.
